WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,262 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include <tilegen/CartesianPower.h>
+
+class TestCartesianPower : public ::testing::Test
+{
+protected:
+    typedef std::vector<int> Container;
+    typedef std::vector<int>::const_iterator Iterator;
+
+    Container base{4, 5, 6};
+    const tilegen::CartesianPower<Iterator> cp{base, 4};
+    TestCartesianPower()
+    {
+
+    };
+    ~TestCartesianPower() {};
+};
+
+TEST_F(TestCartesianPower, TestCartesianPowerSize)
+{
+    EXPECT_EQ(81, cp.size()) <<
+        "CartesianProduct has incorrect size";
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCount)
+{
+    size_t count = 0;
+    for (const auto& it : cp)
+        count++;
+    EXPECT_EQ(81, count);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonBeginEnd)
+{
+    EXPECT_EQ(cp.cbegin(), cp.cbegin());
+    EXPECT_NE(cp.cend(), cp.cbegin());
+    EXPECT_NE(cp.cbegin(), cp.cend());
+    EXPECT_EQ(cp.cend(), cp.cend());
```

Comment:
They produce different function calls: `cend().operator==(cbegin())` and `cbegin().operator==(cend())`. `CartesianPowerIterator`s with `isEnd() == true` are quite different internally from dereferenceable `CartesianPowerIterator`s, so it's worthwhile to check that comparison operators work correctly no matter which instance is `*this`.